### PR TITLE
Rehome libcxx-builder docker image & attempt gentler termination.

### PR DESCRIPTION
**Pull request by EricWF**
There are three changes present in this PR.

1. Use github packages for libcxx-builder rather than dockerhub.

The ldionne/libcxx-builder image will now be hosted at ghcr.io/libcxx/libcxx-builder. This has the benefit of allowing members of the github org to push new versions.

In the future I hope to add github actions to rebuild the image as needed.

2. Add docker-compose file

The compose file allows to to specify the package repository, so that users can simply write 'docker compose build' and 'docker compose push'.

It also gives us a centralized place to manage version arguments which change frequently.

3. Use non-shell CMD form.

This may help the google libcxx builders disconnect more gracefully as the shell form of CMD may eat the shutdown signal. I'm hoping this corrects inaccurate agent counts from the buildkite API, since when the VM's terminate, they do so without signaling it to buildkite, which hangs around waiting for them to reconnect. It's likely more changes will be needed though.

### CODE REVIEW

**Comment by EricWF** [comment]
FYI I cancelled the pre-commit on this change, since it doesn't affect anything the pre-commit tests.

File: libcxx/utils/ci/Dockerfile
Comment (mordante) [inline libcxx/utils/ci/Dockerfile:68]:
```suggestion
ARG LLVM_HEAD_VERSION # populated in the docker-compose file
```

File: libcxx/utils/ci/Dockerfile
Comment (mordante) [inline libcxx/utils/ci/Dockerfile:28]: There is "docker-compose" and "docker compose"
On Ubuntu 22.04 (the lastest LTS version)  "docker compose" is not available.
This link contains more information.
https://github.com/docker/compose-cli/issues/1404#issue-823883873

Can you add a comment that "docker-compose" can be used when "docker compose" is not available?

**Review by mordante — APPROVED** [review]
Thanks for working on this.  LGTM modulo some comments. Please wait for @ldionne to approve too.

File: libcxx/utils/ci/Dockerfile
Comment (EricWF) [reply libcxx/utils/ci/Dockerfile:28]: I don't know that it can be, I can't test. And I think we want to require modern enough versions of docker to use modern features.

You can get docker compose on Ubuntu using the official install scripts. [1] It's how I get docker on debian. And while using additional package repos is normally smelly, it seems to be the norm for docker.

```
curl -fsSL https://get.docker.com -o get-docker.sh
sudo sh get-docker.sh
```

[1] https://docs.docker.com/engine/install/ubuntu/

**Review by ldionne — APPROVED** [review]
Thanks for doing this! Are the permissions on ghcr.io/libcxx set up already so that we can start building the image and pushing it?

File: libcxx/utils/ci/docker-compose.yml
Comment (tstellar) [inline libcxx/utils/ci/docker-compose.yml:4]: Did you create a libcxx organization for this?  Is there any reason not to use the llvm org?

**Comment by EricWF** [comment]
> Thanks for doing this! Are the permissions on ghcr.io/libcxx set up already so that we can start building the image and pushing it?

If you're a member of the github.com/libcxx org, you should be able to `docker login` to the ghcr by following these instructions: 

https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-with-a-personal-access-token-classic

It's probably worth having someone else test this to confirm it works.

File: libcxx/utils/ci/docker-compose.yml
Comment (EricWF) [reply libcxx/utils/ci/docker-compose.yml:4]: I've had the org for years now after noticing it was unclaimed. I recently started using it to test and prototype self hosted github actions. It's a lot easier to iterate the much smaller org.

Github permissions still seem rather course grained, which might be an issue.

Comment (tstellar) [reply libcxx/utils/ci/docker-compose.yml:4]: OK, I really like the idea of using the github container registry for this, but I'm going to recommend you use the llvm org to host the packages for a few reasons:

- In the past the LLVM Foundation has discouraged sub-projects from creating their own spaces on GitHub separate from the main project.  Having a libcxx org for testing and prototype doesn't seem like a problem to me, but I think hosting official project images there may cross the line.
- Bandwidth charges: GitHub has recently extended it's free period for container storage and bandwidth, so this isn't an issue today, but in the future if they start charging, you will likely need to request a budget from the LLVM Foundation for this which would require moving the containers to the llvm org.  Also, you aren't charged for bandwidth when downloads come from GitHub Actions, but this only applies to containers in your current org from what I can tell.

I understand that it's easier to manage permissions using the libcxx org, but it sounds like there be specific fine-grained permissions just for the container registry, so I can look into setting that up for you.

Comment (EricWF) [reply libcxx/utils/ci/docker-compose.yml:4]: I don't disagree with anything you have said.

The libc++ org is better than the status quo of ldionne's personal dockerhub, but LLVM's org is the most appropriate.

Comment (ldionne) [reply libcxx/utils/ci/docker-compose.yml:4]: @tstellar One thing we're also trying to avoid is being too dependent on heavy processes or bottlenecks for making modifications of this kind. Right now with how permissions are set up in the LLVM org, we're not empowered to make these kinds of changes IIUC. With our own organization, we are.

I am not disagreeing that LLVM is the right place to do all of it (it totally is), but in the balance is also our ability to actually get stuff working, maintain them and solve our problems without having to wait for someone with enough admin privileges to set things up for us (even though you all have been amazingly responsive and helpful whenever we needed help). My request for admin-ships in the LLVM org was related to this amongst other things.

**Comment by EricWF** [comment]
I'm going to go ahead and merge this, and we'll move it to LLVM once the permission issues are cleared up. Merging this unblocks me on other work.

**Pull request merged**
